### PR TITLE
Rework related to time api stuffs

### PR DESCRIPTION
**Pull request by pan3793**
This PR based on #275 with reworks and additional changes .

Basically, the changes are:

1. change `DataTypeDate` serde type from `java.sql.Date` to `java.time.LocalDate`
2. change `DataTypeDateTime` and `DataTypeDateTime64` serde type from `java.sql.Timestamp` to `java.time.ZonedDateTime`
3. introduce `IDataType#jdbcJavaType`
4. implement `ClickHouseResultSet.getTimestamp(col, calendar)`

### CODE REVIEW

**Comment by pan3793** [comment]
Benchmark Report
===
```
Benchmark                                           (batchSize)  (columnNum)  (selectNumber)  Mode  Cnt      Score   Error  Units
InsertIBenchmark.benchInsertHttp                         200000           20             N/A  avgt         346.257          ms/op
InsertIBenchmark.benchInsertHttp                         200000           50             N/A  avgt         331.623          ms/op
InsertIBenchmark.benchInsertHttp                         500000           20             N/A  avgt         909.418          ms/op
InsertIBenchmark.benchInsertHttp                         500000           50             N/A  avgt         880.718          ms/op
InsertIBenchmark.benchInsertNative                       200000           20             N/A  avgt          85.350          ms/op
InsertIBenchmark.benchInsertNative                       200000           50             N/A  avgt         100.374          ms/op
InsertIBenchmark.benchInsertNative                       500000           20             N/A  avgt         203.551          ms/op
InsertIBenchmark.benchInsertNative                       500000           50             N/A  avgt         227.984          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       200000           20             N/A  avgt         210.555          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       200000           50             N/A  avgt         510.901          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       500000           20             N/A  avgt         466.658          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       500000           50             N/A  avgt        1150.064          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              200000           20             N/A  avgt         346.747          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              200000           50             N/A  avgt         970.497          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              500000           20             N/A  avgt         861.119          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              500000           50             N/A  avgt        2363.464          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          200000           20             N/A  avgt         171.204          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          200000           50             N/A  avgt         447.883          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          500000           20             N/A  avgt         392.691          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          500000           50             N/A  avgt        1126.907          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 200000           20             N/A  avgt         192.714          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 200000           50             N/A  avgt         509.528          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 500000           20             N/A  avgt         563.014          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 500000           50             N/A  avgt        1494.016          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       200000           20             N/A  avgt         430.095          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       200000           50             N/A  avgt        1069.803          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       500000           20             N/A  avgt        1043.036          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       500000           50             N/A  avgt        2683.945          ms/op
RowBinaryStringIBenchmark.benchInsertNative              200000           20             N/A  avgt         495.865          ms/op
RowBinaryStringIBenchmark.benchInsertNative              200000           50             N/A  avgt        1251.755          ms/op
RowBinaryStringIBenchmark.benchInsertNative              500000           20             N/A  avgt        1192.463          ms/op
RowBinaryStringIBenchmark.benchInsertNative              500000           50             N/A  avgt        2990.398          ms/op
SelectIBenchmark.benchSelectHTTP                            N/A          N/A          500000  avgt        1066.622          ms/op
SelectIBenchmark.benchSelectHTTP                            N/A          N/A        10000000  avgt       22343.706          ms/op
SelectIBenchmark.benchSelectNative                          N/A          N/A          500000  avgt         296.265          ms/op
SelectIBenchmark.benchSelectNative                          N/A          N/A        10000000  avgt        5490.880          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         200000           20             N/A  avgt         770.268          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         200000           50             N/A  avgt        2028.655          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         500000           20             N/A  avgt        1987.401          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         500000           50             N/A  avgt        5087.456          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       200000           20             N/A  avgt         346.697          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       200000           50             N/A  avgt         974.440          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       500000           20             N/A  avgt         848.170          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       500000           50             N/A  avgt        2301.318          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            200000           20             N/A  avgt         397.156          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            200000           50             N/A  avgt        1004.419          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            500000           20             N/A  avgt        1031.446          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            500000           50             N/A  avgt        2470.049          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          200000           20             N/A  avgt         217.321          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          200000           50             N/A  avgt         548.596          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          500000           20             N/A  avgt         568.266          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          500000           50             N/A  avgt        1276.602          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         200000           20             N/A  avgt         476.270          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         200000           50             N/A  avgt        1177.337          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         500000           20             N/A  avgt        1236.201          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         500000           50             N/A  avgt        2967.486          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       200000           20             N/A  avgt         473.742          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       200000           50             N/A  avgt        1280.912          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       500000           20             N/A  avgt        1233.368          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       500000           50             N/A  avgt        3006.721          ms/op
```

**Comment by pan3793** [comment]
Benchmark Report
===
```
Benchmark                                           (batchSize)  (columnNum)  (selectNumber)  Mode  Cnt      Score   Error  Units
InsertIBenchmark.benchInsertHttp                         200000           20             N/A  avgt         429.795          ms/op
InsertIBenchmark.benchInsertHttp                         200000           50             N/A  avgt         432.109          ms/op
InsertIBenchmark.benchInsertHttp                         500000           20             N/A  avgt        1142.507          ms/op
InsertIBenchmark.benchInsertHttp                         500000           50             N/A  avgt        1123.495          ms/op
InsertIBenchmark.benchInsertNative                       200000           20             N/A  avgt         116.759          ms/op
InsertIBenchmark.benchInsertNative                       200000           50             N/A  avgt         111.841          ms/op
InsertIBenchmark.benchInsertNative                       500000           20             N/A  avgt         255.187          ms/op
InsertIBenchmark.benchInsertNative                       500000           50             N/A  avgt         261.087          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       200000           20             N/A  avgt         253.290          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       200000           50             N/A  avgt         616.604          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       500000           20             N/A  avgt         595.582          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       500000           50             N/A  avgt        1448.429          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              200000           20             N/A  avgt         439.747          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              200000           50             N/A  avgt        1372.007          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              500000           20             N/A  avgt        1071.115          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              500000           50             N/A  avgt        3045.541          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          200000           20             N/A  avgt         215.561          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          200000           50             N/A  avgt         537.784          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          500000           20             N/A  avgt         506.868          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          500000           50             N/A  avgt        1545.142          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 200000           20             N/A  avgt         283.146          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 200000           50             N/A  avgt         705.697          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 500000           20             N/A  avgt         731.561          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 500000           50             N/A  avgt        1843.517          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       200000           20             N/A  avgt         551.651          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       200000           50             N/A  avgt        1380.435          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       500000           20             N/A  avgt        1334.493          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       500000           50             N/A  avgt        3128.346          ms/op
RowBinaryStringIBenchmark.benchInsertNative              200000           20             N/A  avgt         605.772          ms/op
RowBinaryStringIBenchmark.benchInsertNative              200000           50             N/A  avgt        1682.322          ms/op
RowBinaryStringIBenchmark.benchInsertNative              500000           20             N/A  avgt        1578.742          ms/op
RowBinaryStringIBenchmark.benchInsertNative              500000           50             N/A  avgt        4022.074          ms/op
SelectIBenchmark.benchSelectHTTP                            N/A          N/A          500000  avgt        1362.302          ms/op
SelectIBenchmark.benchSelectHTTP                            N/A          N/A        10000000  avgt       27999.584          ms/op
SelectIBenchmark.benchSelectNative                          N/A          N/A          500000  avgt         358.990          ms/op
SelectIBenchmark.benchSelectNative                          N/A          N/A        10000000  avgt        6523.757          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         200000           20             N/A  avgt         993.041          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         200000           50             N/A  avgt        2581.496          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         500000           20             N/A  avgt        2455.393          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         500000           50             N/A  avgt        6308.220          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       200000           20             N/A  avgt         429.577          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       200000           50             N/A  avgt        1328.585          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       500000           20             N/A  avgt        1067.372          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       500000           50             N/A  avgt        2997.584          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            200000           20             N/A  avgt         501.217          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            200000           50             N/A  avgt        1266.796          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            500000           20             N/A  avgt        1295.986          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            500000           50             N/A  avgt        3083.849          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          200000           20             N/A  avgt         277.116          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          200000           50             N/A  avgt         625.877          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          500000           20             N/A  avgt         713.301          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          500000           50             N/A  avgt        1773.218          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         200000           20             N/A  avgt         607.558          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         200000           50             N/A  avgt        1554.005          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         500000           20             N/A  avgt        1562.661          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         500000           50             N/A  avgt        3782.701          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       200000           20             N/A  avgt         642.406          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       200000           50             N/A  avgt        1695.960          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       500000           20             N/A  avgt        1536.669          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       500000           50             N/A  avgt        3707.264          ms/op
```

**Comment by pan3793** [comment]
ping @sbouchex to review

**Comment by pan3793** [comment]
Benchmark Report
===
```
Benchmark                                           (batchSize)  (columnNum)  (selectNumber)  Mode  Cnt      Score   Error  Units
InsertIBenchmark.benchInsertHttp                         200000           20             N/A  avgt         395.737          ms/op
InsertIBenchmark.benchInsertHttp                         200000           50             N/A  avgt         411.547          ms/op
InsertIBenchmark.benchInsertHttp                         500000           20             N/A  avgt        1043.485          ms/op
InsertIBenchmark.benchInsertHttp                         500000           50             N/A  avgt        1027.824          ms/op
InsertIBenchmark.benchInsertNative                       200000           20             N/A  avgt         107.735          ms/op
InsertIBenchmark.benchInsertNative                       200000           50             N/A  avgt         115.074          ms/op
InsertIBenchmark.benchInsertNative                       500000           20             N/A  avgt         225.538          ms/op
InsertIBenchmark.benchInsertNative                       500000           50             N/A  avgt         238.820          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       200000           20             N/A  avgt         248.899          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       200000           50             N/A  avgt         687.611          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       500000           20             N/A  avgt         603.479          ms/op
RowBinaryDoubleIBenchmark.benchInsertHttpRowBinary       500000           50             N/A  avgt        1572.457          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              200000           20             N/A  avgt         397.405          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              200000           50             N/A  avgt        1140.008          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              500000           20             N/A  avgt         940.444          ms/op
RowBinaryDoubleIBenchmark.benchInsertNative              500000           50             N/A  avgt        2734.801          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          200000           20             N/A  avgt         202.037          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          200000           50             N/A  avgt         534.930          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          500000           20             N/A  avgt         465.677          ms/op
RowBinaryIntIBenchmark.benchInsertHttpRowBinary          500000           50             N/A  avgt        1752.555          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 200000           20             N/A  avgt         248.993          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 200000           50             N/A  avgt         620.303          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 500000           20             N/A  avgt         628.298          ms/op
RowBinaryIntIBenchmark.benchInsertNative                 500000           50             N/A  avgt        1536.490          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       200000           20             N/A  avgt         492.070          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       200000           50             N/A  avgt        1280.911          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       500000           20             N/A  avgt        1201.145          ms/op
RowBinaryStringIBenchmark.benchInsertHttpRowBinary       500000           50             N/A  avgt        2932.672          ms/op
RowBinaryStringIBenchmark.benchInsertNative              200000           20             N/A  avgt         588.631          ms/op
RowBinaryStringIBenchmark.benchInsertNative              200000           50             N/A  avgt        1530.649          ms/op
RowBinaryStringIBenchmark.benchInsertNative              500000           20             N/A  avgt        1405.214          ms/op
RowBinaryStringIBenchmark.benchInsertNative              500000           50             N/A  avgt        3594.341          ms/op
SelectIBenchmark.benchSelectHTTP                            N/A          N/A          500000  avgt        1289.110          ms/op
SelectIBenchmark.benchSelectHTTP                            N/A          N/A        10000000  avgt       26344.929          ms/op
SelectIBenchmark.benchSelectNative                          N/A          N/A          500000  avgt         341.474          ms/op
SelectIBenchmark.benchSelectNative                          N/A          N/A        10000000  avgt        5944.290          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         200000           20             N/A  avgt         889.361          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         200000           50             N/A  avgt        2376.636          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         500000           20             N/A  avgt        2267.884          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertHttp         500000           50             N/A  avgt        5869.706          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       200000           20             N/A  avgt         383.972          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       200000           50             N/A  avgt        1159.705          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       500000           20             N/A  avgt         932.827          ms/op
WideColumnDoubleInsertIBenchmark.benchInsertNative       500000           50             N/A  avgt        2807.228          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            200000           20             N/A  avgt         459.782          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            200000           50             N/A  avgt        1191.680          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            500000           20             N/A  avgt        1199.096          ms/op
WideColumnIntInsertIBenchmark.benchInsertHttp            500000           50             N/A  avgt        3004.035          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          200000           20             N/A  avgt         259.062          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          200000           50             N/A  avgt         622.153          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          500000           20             N/A  avgt         622.756          ms/op
WideColumnIntInsertIBenchmark.benchInsertNative          500000           50             N/A  avgt        1652.375          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         200000           20             N/A  avgt         602.843          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         200000           50             N/A  avgt        1476.522          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         500000           20             N/A  avgt        1562.074          ms/op
WideColumnStringInsertIBenchmark.benchInsertHttp         500000           50             N/A  avgt        3777.205          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       200000           20             N/A  avgt         614.522          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       200000           50             N/A  avgt        1672.548          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       500000           20             N/A  avgt        1468.636          ms/op
WideColumnStringInsertIBenchmark.benchInsertNative       500000           50             N/A  avgt        3816.731          ms/op
```

File: clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/statement/AbstractPreparedStatement.java [hunk 1]
```diff
@@ -133,6 +147,17 @@ public void setObject(int index, Object x, int targetSqlType, int scaleOrLength)
         setObject(index, x);
     }
 
+    protected Object convertObjectIfNecessary(Object obj) {
+        Object result = obj;
+        if (obj instanceof Date) {
```
[the remark this answers is not in the document]
Comment (pan3793) [reply clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/statement/AbstractPreparedStatement.java:152]: Emm.. know it, but I have't found another way to handle it properly, and the overhead seems acceptable. Ref: 
https://stackoverflow.com/questions/4973359/does-instanceof-operator-generate-a-lot-of-overhead-why

[the remark this answers is not in the document]
Comment (sundy-li) [reply clickhouse-native-jdbc/src/main/java/com/github/housepower/jdbc/statement/AbstractPreparedStatement.java:152]: Ok, seems the performance of this pr is stable.

**Pull request merged**